File: web/lib/finance/currency.ts
import { CURRENCY_CONFIG, type CurrencyKey } from './money'

/**
 * Ordered list of supported currencies — matches the iOS `Currency` enum order.
 * Money is always stored as USD cents; display conversion happens at render time.
 */
export const CURRENCIES: CurrencyKey[] = ['usd', 'cad', 'gbp', 'eur', 'jpy', 'cny', 'bdt']

export const CURRENCY_NAMES: Record<CurrencyKey, string> = {
  usd: 'US Dollar',
  cad: 'Canadian Dollar',
  gbp: 'British Pound',
  eur: 'Euro',
  jpy: 'Japanese Yen',
  cny: 'Chinese Yuan',
  bdt: 'Bangladeshi Taka',
}

/** 1 USD = N units of currency. Used when live FX is unavailable. */
export const FALLBACK_RATE_FROM_USD: Record<CurrencyKey, number> = {
  usd: 1,
  cad: 1.35,
  gbp: 0.78,
  eur: 0.92,
  jpy: 150,
  cny: 7.2,
  bdt: 110,
}

export function currencyCode(key: CurrencyKey): string {
  return CURRENCY_CONFIG[key].code
}

export function currencySymbol(key: CurrencyKey): string {
  return CURRENCY_CONFIG[key].symbol
}

export function fractionDigits(key: CurrencyKey): number {
  return CURRENCY_CONFIG[key].fractionDigits
}

export type { CurrencyKey }

File: iOS/Ortho-iOS/Models/Currency.swift
import Foundation

/// User-selectable display currency. Amounts are always stored as USD cents
/// (`Int64`) on `Transaction.amount`; conversion to this currency happens at
/// every render via `Money.string(cents:currency:rate:)` and
/// `AppState.formatMoney(_:)`.
enum Currency: String, CaseIterable, Identifiable, Hashable, Codable {
    case usd, cad, gbp, eur, jpy, cny, bdt

    var id: String { rawValue }

    /// ISO 4217 code — used as `NumberFormatter.currencyCode` for grouping +
    /// fraction digits per locale.
    var code: String { rawValue.uppercased() }

    /// The currency's display symbol. A FIXED table (NOT locale-derived) so it
    /// is byte-identical to web `CURRENCY_CONFIG[...].symbol` and never drifts
    /// with the device locale — in particular CNY stays "CN¥" (disambiguated
    /// from JPY "¥") instead of collapsing to a locale-chosen "¥".
    /// Locked by `shared/test-vectors/currency-symbols.json`.
    var symbol: String {
        switch self {
        case .usd: "$"
        case .cad: "CA$"
        case .gbp: "£"
        case .eur: "€"
        case .jpy: "¥"
        case .cny: "CN¥"
        case .bdt: "৳"
        }
    }

    var displayName: LocalizedStringResource {
        switch self {
        case .usd: "US Dollar"
        case .cad: "Canadian Dollar"
        case .gbp: "UK Pound"
        case .eur: "Euro"
        case .jpy: "Japanese Yen"
        case .cny: "Chinese Yuan"
        case .bdt: "Bangladeshi Taka"
        }
    }

    /// 2 for most currencies, 0 for those with no minor unit (JPY).
    /// Used by amount-field formatting and `Money.toDisplayAmount`.
    var fractionDigits: Int {
        switch self {
        case .jpy: 0
        default:   2
        }
    }

    /// Approximate rate (1 USD = N of this currency) used until a live
    /// fetch lands or when the network is unreachable. Live rates from
    /// floatrates.com override these via `AppState.fxRates`.
    var fallbackRateFromUSD: Decimal {
        switch self {
        case .usd: 1
        case .cad: 1.35
        case .gbp: 0.78
        case .eur: 0.92
        case .jpy: 150
        case .cny: 7.20
        case .bdt: 110
        }
    }
}
